### ISSUE TITLE
$? doesn't change in case of parsing error

### ISSUE THREAD

**Issue by dronkoff** [PowerShell/PowerShell#3798]
Steps to reproduce
------------------
```powershell
1..5 | buggy_call
$?
```
The result is $False and this is expected

```powershell
1..5 | $_
$?
```
The result is $True and this is not expected

Expected behavior
-----------------
```powershell
1..5 | $_
$?
```
The result of $? should be $False because previous statement results in an error.

Environment data
----------------

```powershell
PS C:\temp> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.15063.296
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.15063.296
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```
Notes
----------------
From Tom Nolan words:

> The problem here is that $? Is set during the CheckActionPreference method when an error is encountered.  https://github.com/PowerShell/PowerShell/blob/02b5f357a20e6dee9f8e60e3adb9025be3c94490/src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs
> 
> However, CheckActionPreference is not executed until the AST has been parsed (https://github.com/PowerShell/PowerShell/blob/02b5f357a20e6dee9f8e60e3adb9025be3c94490/src/System.Management.Automation/engine/parser/Compiler.cs) and PowerShell actually attempts to execute the command, stepping through the pipeline.

**Comment by mklement0** [comment]
Good catch, but it's worth noting that parsing errors _by themselves_ DO set `$?` to `$False` (`1 / 0; $?` yields `$False`). 

The problem  you describe surfaces _in pipeline segments other than the 1st_.

Therefore, could you please amend the title of this issue accordingly, and also fix the typo in it (`&?` should be `$?`)?

Also, where is the Tom Nolan quote from?

**Comment by MaximoTrinidad** [comment]
At the same, this issue can be replicated in PowerShell Core (v6 Beta).

![posh_issue_2017-05-17_15-45-56](https://cloud.githubusercontent.com/assets/19892435/26172902/f99ccdf6-3b17-11e7-8267-86379ece3d04.jpg)

Please, I may suggest to also post this incident at Windows PowerShell UserVoice site at: https://windowsserver.uservoice.com/forums/301869-powershell/category/148044-powershell-engine

This way the Windows PowerShell team can assist in taking a look at this issue.

:)

**Comment by BrucePay** [comment]
PowerShell compiles the entire text of a script at once so a parse error anywhere in the block prevents the entire block from being executed. QED you can't catch or detect a parse error in the code being parsed since that code is never run. @mklement0 's example of 1/0 is not a parse error. (It was treated as a parse-time error in PowerShell V1 but it's been a runtime error since then.)

**Comment by mklement0** [comment]
Thanks for the clarification, @BrucePay - _parse error_ was definitely the wrong term to use (I've deleted my comment).

I don't fully understand what's going on, but it seems that the error record generated by the broken pipeline segment (`Expressions are only allowed as the first element of a pipeline`) is implicitly passed to the `Out-Default` cmdlet and perhaps it is therefore _`Out-Defaults` success_ that is reflected in `$?`.

It's certainly unexpected behavior and relates to  [#3359](https://github.com/PowerShell/PowerShell/issues/3359#issuecomment-287842045) and whether it wouldn't be better to treat such errors as _terminating_ errors.

**Comment by iSazonov** [comment]
Really in second example a pipeline is created but [interrupted](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs#L367) and [the code](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs#L407) that set QuestionMarkVariableValue don't executed. 
Default value for QuestionMarkVariableValue is True - so we see True as result.

I believe it is by design - (1) as @BrucePay said it is not problem for script because _no command_ will be executed - that's why we should use try-catch. (2) for interactive session using $? is pointless while you see parser error - I believe we shouldn't worry about QuestionMarkVariableValue value in that case.

**Comment by mklement0** [comment]
@iSazonov: My bad (I have little insight into how PS parsing works at this point; other comment deleted too):

I missed that if the code is placed in a script, the _entire script_ won't be executed, so it _is_ treated as  parse error. 

Interactively, this can be shown by placing `$?` on the _same_ line, in which case `$?` is never executed:

```powershell
1..5 | $_; $?   # $? never executes
```

Calling a script that fails to parse _from another script_ DOES set `$?` to `$False`, so the interactive behavior - though unexpected - is less of an issue, as you say.

On a related note, however: from the calling script's perspective, shouldn't an attempt to call a script _with parse errors_ result in a _terminating_ error that the calling script must therefore handle with `try` / `catch`?

**Comment by iSazonov** [comment]
@mklement0 Please don't remove history. You can strike out - use ~~double tildes around the words~~.

**Comment by iSazonov** [comment]
My understanding is that $? is from v1 time and today prefered way is ` try  /  catch`

**Comment by mklement0** [comment]
Good point re history, will do.

`$?` still has (and, I presume, will continue to have) its place for reflecting _non-terminating_ errors, but a parse error is arguably severe enough to warrant a terminating one so that the script aborts _by default_.

Yes, you _can_ detect the parse error with `try` / `catch` (the exception inside the error record is of type `[System.Management.Automation.ParseException]`), but _by default_ the calling script continues to run.

Note that an ordinary non-terminating error - such as reported by `Get-Item /Nosuch` can _not_ be caught with `try` / `catch`.

From what I can tell, this **problematic hybrid behavior** - non-terminating-error behavior by default, terminating-error behavior with `try` / `catch` - applies to the following kinds of errors:

 * parse errors in an invoked script (`[System.Management.Automation.ParseException]`)
 * runtime errors such as `1 / 0` (`[System.Management.Automation.RuntimeException]`)
 * .NET methods that report an exception, such as`[int]::Parse('foo')` (`[System.Management.Automation.MethodInvocationException]`)

Note sure if there are more.

As I've suggested in https://github.com/PowerShell/PowerShell/issues/3359#issuecomment-287842045, **perhaps it makes more sense to treat these as _terminating_ errors.**

**Comment by iSazonov** [comment]
In PowerShell context "terminating" error is a error "terminating current pipeline".
As previously mentioned, we have confusion with terminating and non-terminating errors.

**Comment by rkeithhill** [comment]
> My understanding is that $? is from v1 time and today a prefered way is try / catch

`$?` is still useful in scripts to see if a command encountered a non-terminating error.  Say I'm copying a bunch of files recursively from one dir to another.  If there is a file or folder I don't have permission to the copy will not 100% succeed.  I can test for this condition by checking `$?`.  

Of course, you can tell PowerShell to convert the non-terminating error to a terminating error with `-ErrorAction Stop` and then use try/catch.  Which one a scripter chooses comes down to preference - what a scripter is more comfortable with.

**Comment by mklement0** [comment]
@rkeithhill: 

Absolutely, but my point was that the error types I've mentioned (parse in callee/runtime/parameter binding/method invocation):

* behave like _non-terminating_ errors by default
* though in their ability to be caught _irrespective of the current `$ErrorActionPreference` value_ they behave like _terminating ones_ - this is what I meant by "hybrid".

It seems to me that making these error types regular _terminating_ errors not only resolves the inconsistency, but also makes more sense:
These types of errors are _not_ the ones you want to _tolerate_ as _partial failures_ while the script goes on to still perform partially useful things.

For instance, if I call `Get-Content -NoSuchParameter` in a script, I want that script to _break_ by default. 

With respect to the general distinction between terminating and non-terminating errors and to what extent PS and its cmdlets play by the official rules, I've tried to summarize the state of affairs in [this SO answer](http://stackoverflow.com/a/39949027/45375) a while back.

**Comment by dronkoff** [comment]
>     $?  
>        Contains the execution status of the last operation. It contains
>        TRUE if the last operation succeeded and FALSE if it failed

From the end user point of view it doesn't matter what type of error he sees on the screen. He sees red text and an error description but $? returns $true. He decides "ok, look like this thing doesn't work, I wont use it" and then he goes to: 
```powershell
$error.Clear()
... -ErrorAction SilentlyContinue
if($error.Count -gt 0) 
```
Looks like it would be great to change the default value to $false.
If no operation was executed then how status of it's execution can be $true.

**Comment by lzybkr** [comment]
I do question the value of making a change here. `$?` is most useful when it's not obvious if there was an error or not, and it's especially useful in automation.

Parse errors that occur during automation should be exceedingly rare - it basically means you didn't test your script before deployment.

**Comment by mklement0** [comment]
@lzybkr: I realize I've hijacked this thread a bit.

* _Interactively_, I do see that the current behavior is not too problematic, despite its awkwardness.

* When it comes to _automation_:

  * Even with respect to parse errors alone I would want a script that calls a non-parseable script to _fail right away_, even  _while developing_ a script.

  * There are also the less predictable cases I've mentioned (runtime/parameter binding/method invocation), where a script _breaking_ would be much more helpful than just continuing.

Just so we have clarity:

* Is the only reason for not wanting to make a change _backward compatibility_?

* Does this apply to _all_ of the error types mentioned, or is it worth opening a separate issue for at least _some_ of them?

**Comment by BrucePay** [comment]
@mklement0 

> Even with respect to parse errors alone I would want a script that calls a non-parseable script to fail right away, even while developing a script.

Command resolution is done at run time (commands are late-bound). A called script won't be compiled until it is called. It may not even be called at all depending on the runtime logic of the calling script.

**Comment by mklement0** [comment]
@BrucePay:

Yes, but that's not what I meant. I was responding to the following comment by @lzybkr:

>  it basically means you didn't test your script before deployment.

What I was trying to say was: _at runtime_, an attempt to call a script that fails to parse should cause a _script_-terminating error (runspace-terminating error), not a _statement_-terminating error - whether in production or during development.

It simply makes no sense to continue execution _by default_ in such a scenario, which is what currently happens.

**Comment by BrucePay** [comment]
@mklement0 I get that parse errors _feel_ different, but thinking practically, why should a parse error be handled differently from any other fatal error that the script might encounter? Either way, from the caller's perspective, it's just another command that failed.  Now you could argue that **all** unhandled errors should be fatal and terminate the script by default. That's what C# effectively does. But that's not what shells (bash, cmd, etc.) do and PowerShell is a shell. But PowerShell is also a scripting language and so we have hybrid semantics with exit status ($?) and exceptions. In fact part of the reason for deciding to have the `trap` statement was that it made it easy to change how fatal errors are handled throughout a script. 

(Small note: you probably don't want syntax errors to be runspace terminating otherwise your interactive session would end the first time you mistyped a command. That seems undesirable :-)

**Comment by mklement0** [comment]
@BrucePay  I'll address the conceptual issues later; but re your small note:

> you probably don't want syntax errors to be runspace terminating otherwise your interactive session would end the first time you mistyped a command.

Runspace-terminating errors don't terminate interactive sessions; I don't know how interactive sessions are implemented, but executing something like `Throw "a fit"` on the command line does _not_ terminate the session.

**Comment by BrucePay** [comment]
@mklement0 
>>    Now you could argue that all unhandled errors should be fatal and terminate the script by default.

> Indeed, that's exactly what should happen

Think about that for a second. There are a lot of little errors where people just let the error message go to the output stream, Instead you want these errors to terminate the entire thread of execution? If nothing else, it would make the interactive experience unusable.

> Runspace-terminating errors don't terminate interactive sessions; I don't know how interactive sessions are implemented, but executing something like Throw "a fit" on the command line does not terminate the session.

_Runspace terminating errors most certainly terminate your session_ since a session is nothing but a wrapper around a runspace. Your example
```
Throw "a fit" 
```
is not a runspace terminating exception. It's just an exception that will, if uncaught, terminate the current thread of execution. The runspace, however, will remain open (i.e. not terminated). If you do want to try terminating you runspace, fell free to try
```
[runspace]::defaultrunspace.close()
```

**Comment by mklement0** [comment]
@BrucePay:

>  It's just an exception that will, if uncaught, terminate the current thread of execution. 

In the absence of information in the documentation, I've been struggling to come up with the right terminology to describe the unit of execution that gets terminated when you execute something like `Throw "a fit"` - I've called it a _script-terminating_ error and then, sensing it might not be the best term, _guessed_ at _runspace-terminating_ error.

Based on what you're saying, the most accurate term is therefore _thread-terminating_ error?
Until I have confirmation, I'll call them _fatal errors_ - which may be the more intuitive term anyway (which would still leaves us with the potentially misleading term _terminating error_ in the existing documentation,, where it more often than not means _statement_-terminating error).
Do tell us what you think the term should be

---

Therefore, what I was really advocating was to treat cases such as `1 / 0` and `[int]::Parse('foo')` the same as `Throw "a fit"`, i.e., to treat them as _fatal errors_ (by default).

In other words: yes, inside a script, I want that script to terminate with an error by default in all these cases.

**Comment by BrucePay** [comment]
>  I've called it a script-terminating error and then, sensing it might not be the best term,

Script-terminating is generally fine. Or "terminates the current thread of execution".

> Therefore, what I was really advocating was to treat cases such as 1 / 0 and [int]::Parse('foo') the same as Throw "a fit", i.e., to treat them as fatal errors (by default).

That's easy - put a trap at the top level of your script:
```
trap { break }
& {
    &{
        foreach ($i in 0..5)
        {
            1/$i
            "Hello"
        }
    }
}
```
This will turn statement-terminating error into a "fatal" one.

**Comment by mklement0** [comment]
@BrucePay:

> Script-terminating is generally fine. Or "terminates the current thread of execution".

Thanks. I've since kind of fallen in love with _fatal error_ for its brevity, though (a fatal attraction, if you will), so it may leak out from time to tome. Let me know if you think that's not appropriate.

(My original rationale for coining the term _script-terminating_ was to (a) use a term for the unit of execution being terminated familiar to even casual PowerShell users and (b) to keep the word "terminating" in there so as tease out the difference between a _statement_-terminating and a _thread_-terminating error so conspicuously absent from the documentation. It doesn't tell the fully story, however - but perhaps it doesn't need to.)

Yes, your `trap { break }` solution effectively does what I'm asking for, but my point is that this should always have been the _default behavior_ - but I know that that only matters in the context of the "PowerShell vZeroTechnicalDebt" conversation.
